src/server/raft_client.rs

```diff
@@ -81,41 +90,44 @@ impl RaftClient {
         }
     }
 
-    fn get_conn(&mut self, addr: SocketAddr, index: usize) -> &mut Conn {
+    fn get_conn(&mut self, addr: SocketAddr, region_id: u64, store_id: u64) -> &mut Conn {
         let env = self.env.clone();
         let cfg = self.cfg.clone();
+        let index = region_id as usize % cfg.grpc_raft_conn_num;
         self.conns
             .entry((addr, index))
-            .or_insert_with(|| Conn::new(env, addr, &cfg))
+            .or_insert_with(|| Conn::new(env, addr, &cfg, store_id))
     }
 
     pub fn send(&mut self, store_id: u64, addr: SocketAddr, msg: RaftMessage) -> Result<()> {
-        let index = msg.get_region_id() as usize % self.cfg.grpc_raft_conn_num;
-        let res = {
-            let mut conn = self.get_conn(addr, index);
-            conn.active = true;
-            UnboundedSender::send(&conn.stream, (msg, WriteFlags::default().buffer_hint(true)))
-        };
-        if let Err(e) = res {
-            warn!("server: drop conn with tikv endpoint {} error: {:?}",
-                  addr,
-                  e);
-            self.conns.remove(&(addr, index));
-            self.addrs.remove(&store_id);
-            return Err(box_err!(e));
-        }
+        let mut conn = self.get_conn(addr, msg.region_id, store_id);
+        conn.buffer.as_mut().unwrap().push((msg, WriteFlags::default().buffer_hint(true)));
         Ok(())
     }
 
 
     pub fn flush(&mut self) {
-        for conn in self.conns.values_mut() {
-            if conn.active {
-                conn.active = false;
-                if let Err(e) = UnboundedSender::send(&conn.stream,
-                                                      (RaftMessage::new(), WriteFlags::default())) {
-                    error!("flush conn error {:?}", e);
-                }
+        let mut dead_conns = Vec::new();
+        for (key, conn) in &mut self.conns {
+            if conn.buffer.as_ref().unwrap().is_empty() {
+                continue;
+            }
+
+            let mut msgs = conn.buffer.take().unwrap();
+            msgs.last_mut().unwrap().1 = WriteFlags::default();
+            if let Err(e) = UnboundedSender::send(&conn.stream, msgs) {
+                error!("server: drop conn with tikv endpoint {} flush conn error: {:?}",
+                       key.0,
+                       e);
+                dead_conns.push(*key);
+            }
+
+            conn.buffer = Some(Vec::with_capacity(INITIAL_BUFFER_CAP));
+        }
+
+        for key in dead_conns {
+            if let Some(conn) = self.conns.remove(&key) {
+                self.addrs.remove(&conn.store_id);
```

Comment:
There may be some dead connection leaked in HashMap if the new address is different from the old one.